src/cargo/sources/registry/http_remote.rs

```diff
@@ -0,0 +1,1209 @@
+                self.downloads.downloaded_bytes += dl.total.get();
+                self.downloads.tick(WhyTick::DownloadFinished)?;
+
+                match code {
+                    200 => {
+                        // We got data back, hooray!
+                        // Let's update the index file.
+                        let path = self.config.assert_package_cache_locked(&self.index_path);
+                        let pkg = path.join(&fetched.path);
+                        paths::create_dir_all(pkg.parent().expect("pkg is a file"))?;
+                        let mut file = paths::create(pkg)?;
+                        file.write_all(dl.etag.into_inner().as_deref().unwrap_or("\n").as_bytes())?;
+                        file.write_all(
+                            dl.last_modified
+                                .into_inner()
+                                .as_deref()
+                                .unwrap_or("\n")
+                                .as_bytes(),
+                        )?;
+                        file.write_all(&data)?;
+                        file.flush()?;
+
+                        assert!(
+                            self.downloads
+                                .eager
+                                .insert(fetched.path.clone(), fetched)
+                                .is_none(),
+                            "download finished for already-finished path"
+                        );
+                    }
+                    304 => {
+                        // Not Modified response.
+                        // There's nothing for us to do -- the index file is up to date.
+                        // The only thing that matters is telling the caller about this package.
+                        assert!(
+                            self.downloads
+                                .eager
+                                .insert(fetched.path.clone(), fetched)
+                                .is_none(),
+                            "download finished for already-finished path"
+                        );
+                    }
+                    403 | 404 => {
+                        // Not Found response.
+                        // We treat Forbidden as just being another expression for 404
+                        // from a server that does not want to reveal file names.
+                        // The crate doesn't exist, so we simply do not yield it.
+                        // Errors will eventually be yielded by load().
+                    }
+                    410 | 451 => {
+                        // The crate was deleted from the registry.
+                        // Errors will eventually be yielded by load().
+                        todo!("we should delete the local index file here if it exists");
+                    }
+                    code => {
+                        anyhow::bail!(
+                            "prefetch: server returned unexpected HTTP status code {} for {}{}",
```

Comment:
FWIW this error I think will want to get improved to also possibly mention header values and/or part of the response body. Otherwise it's sort of opaque to debug why we got a particular error code from a url.

Reply:
It now includes the first line of the body as well. I'm hesitant to include the full body, as it could be _quite_ verbose. Also, in all likelihood the user can just `curl` the same URL and get the full response.